File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/LineRenderer.kt
package com.nativephp.mobile.ui.nativerender

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.View

class LineViewRenderer : NativeViewRenderer {
    override fun createView(context: Context, node: NativeUINode): View {
        return LineView(context, node)
    }

    override fun updateView(view: View, node: NativeUINode) {
        (view as? LineView)?.updateNode(node)
    }
}

private class LineView(context: Context, private var node: NativeUINode) : View(context) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val density = context.resources.displayMetrics.density

    fun updateNode(newNode: NativeUINode) {
        node = newNode
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        val p = node.props
        val fromX = p.getFloat("from_x", 0f)
        val fromY = p.getFloat("from_y", 0f)
        val toX = p.getFloat("to_x", 0f)
        val toY = p.getFloat("to_y", 0f)

        val style = node.style
        paint.color = if (style != null && style.borderColor != 0) style.borderColor else Color.BLACK
        paint.strokeWidth = if (style != null && style.borderWidth > 0) style.borderWidth * density else density

        canvas.drawLine(
            fromX * density,
            fromY * density,
            toX * density,
            toY * density,
            paint
        )
    }
}

File: resources/xcode/NativePHP/NativeRender/LineRenderer.swift
import UIKit

struct LineViewRenderer: NativeViewRenderer {
    func createView(node: NativeUINode) -> UIView {
        let v = LineView()
        applyProps(v, node: node)
        return v
    }

    func updateView(_ view: UIView, node: NativeUINode) {
        guard let v = view as? LineView else { return }
        applyProps(v, node: node)
        v.setNeedsDisplay()
    }

    private func applyProps(_ v: LineView, node: NativeUINode) {
        let p = node.props
        v.fromPoint = CGPoint(x: CGFloat(p.getFloat("from_x")), y: CGFloat(p.getFloat("from_y")))
        v.toPoint = CGPoint(x: CGFloat(p.getFloat("to_x")), y: CGFloat(p.getFloat("to_y")))
        v.strokeColor = (node.style?.borderColor != 0)
            ? UIColor(argb: node.style!.borderColor)
            : .black
        v.strokeWidth = (node.style?.borderWidth ?? 0) > 0
            ? CGFloat(node.style!.borderWidth)
            : 1.0
        v.backgroundColor = .clear
        v.isOpaque = false
        v.isUserInteractionEnabled = false
    }
}

private class LineView: UIView {
    var fromPoint: CGPoint = .zero
    var toPoint: CGPoint = .zero
    var strokeColor: UIColor = .black
    var strokeWidth: CGFloat = 1.0

    override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        ctx.setStrokeColor(strokeColor.cgColor)
        ctx.setLineWidth(strokeWidth)
        ctx.move(to: fromPoint)
        ctx.addLine(to: toPoint)
        ctx.strokePath()
    }
}
